lib/Optimizer/GraphOptimizer/PassManager.cpp

```diff
@@ -0,0 +1,198 @@
+    llvm::cl::desc("Debug option to export the Graph in DOT format before the "
+                   "listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(passManagerCat));
+
+llvm::cl::opt<bool> dumpDAGAfterAllPassesOpt(
+    "dump-graph-DAG-after-all-passes",
+    llvm::cl::desc(
+        "Debug option to export the Graph in DOT format after all passes."),
+    llvm::cl::Optional, llvm::cl::cat(passManagerCat));
+
+llvm::cl::list<std::string> dumpDAGAfterPassesOpt(
+    "dump-graph-DAG-after-passes",
+    llvm::cl::desc("Debug option to export the Graph in DOT format after the "
+                   "listed Passes."),
+    llvm::cl::value_desc("FunctionPass names (e.g. DCE,CSE)"),
+    llvm::cl::ZeroOrMore, llvm::cl::CommaSeparated,
+    llvm::cl::cat(passManagerCat));
+
+llvm::cl::opt<bool> printPassesOpt(
+    "print-passes",
+    llvm::cl::desc("Print all of the passes run by the pass manager."),
+    llvm::cl::Optional, llvm::cl::cat(passManagerCat));
+
+llvm::cl::opt<unsigned> stopAfterPassNumOpt(
+    "stop-passes-after-num",
+    llvm::cl::desc("Number of passes to run before preventing running any "
+                   "passes. Used for debugging."),
+    llvm::cl::init(std::numeric_limits<unsigned>::max()),
+    llvm::cl::cat(passManagerCat));
+
+/// Helper to check if \p otherStr is in \p strList.
+static bool listContainsString(llvm::ArrayRef<std::string> strList,
+                               llvm::StringRef otherStr) {
+  for (llvm::StringRef str : strList) {
+    if (str == otherStr) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Global pass counter used to identify each pass.
+static unsigned globalPassCounter = 0;
+
+} // namespace
+
+bool FunctionPassManager::runPrePass(Function *F,
+                                     const CompilationContext &cctx,
+                                     const FunctionPass &P) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Starting Pass #" << globalPassCounter << ": "
+              << P.getName().str() << " on Function: \""
+              << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGBeforeAllPassesOpt ||
+      listContainsString(dumpDAGBeforePassesOpt, P.getName())) {
+    F->dumpDAG(F->getName().str() + "_PrePass_" + P.getName().str() + "_n" +
+               std::to_string(globalPassCounter) + ".dot");
+  }
+  if (verifyBeforeAllPassesOpt ||
+      listContainsString(verifyBeforePassesOpt, P.getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+bool FunctionPassManager::runPostPass(Function *F,
+                                      const CompilationContext &cctx,
+                                      const FunctionPass &P) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Finished Pass #" << globalPassCounter << ": "
+              << P.getName().str() << " on Function: \""
+              << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGAfterAllPassesOpt ||
```

Comment:
See my comments about moving this kind of logic into more specialized derived classes.